text/1615-cargo-platform-standards.md

```diff
@@ -0,0 +1,170 @@
+
+* Using a `.cargo` directory is violating the recommendations/rules on most
+  operating systems. _(Linux, Windows, macOS. Especially painful on Windows,
+  where dotfiles are not hidden.)_
+* Putting caches in designated cache directories allows backup tools to ignore
+  them. _(Linux, Windows, macOS. Example: Time Machine ignores the cache
+  directory on macOS.)_
+* It makes it easier for users to manage, share and version-control their
+  configuraton files, as configuration files from different applications end up
+  in the same place, instead of being intermingled with cache files. _(Linux
+  and macOS.)_
+* Cargo contributes to the slow cleanup of the `$HOME` directory by stopping to
+  add its application-private clutter to it. _(Linux.)_
+* Using a standard directory for binary outputs can allow the user to execute
+  Cargo-installed binaries without modifying their `PATH` variable. _(Linux)_
+
+Solving this problem will likely also solve the same problem in Cargo-related
+tools such as `rustup` as their strategy is "do what Cargo does".
+
+This seems to be implemented in pip, the Python package manager, already.
+
+# Detailed design
+
+We are going to introduce new environment variables:
+```
+CARGO_BIN_DIR
+CARGO_CACHE_DIR
+CARGO_CONFIG_DIR
+```
+
+For the default values of these variables if they are set to the empty string
+or not set (which will be the common case), see below.
+
+These will be used to split the current `.cargo` (`CARGO_HOME`) directory up:
+The cached packages (`.cargo/git`, `.cargo/registry`) will go into
+`CARGO_CACHE_DIR`, binaries (`.cargo/bin`) installed by Cargo will go into
+`CARGO_BIN_DIR` and the config (`.cargo/config`) will go into
+`CARGO_CONFIG_DIR`.
+
+In order to maintain backward compatibility, the old directory locations will
+be checked if the new ones don't exist. In detail, this means:
+
+1. If any of the new variables `CARGO_BIN_DIR`, `CARGO_CACHE_DIR`,
+   `CARGO_CONFIG_DIR` are set and nonempty, use the new directory structure.
+2. Else, if there is an override for the legacy Cargo directory, using
+   `CARGO_HOME`, the directories for cache, configuration and executables are
+   placed inside this directory.
+3. Otherwise, if the Cargo-specfic platform-specific directories exist, use
+   them. What constitutes a Cargo-specific directory is laid out below, for
+   each platform.
+4. If that's not the case, check whether the legacy directory exists (`.cargo`)
+   and use it in that case.
+5. If everything else fails, create the platform-specific directories and use
+   them.
+
+This makes Cargo use platform-specific directories for new installs while
+retaining compatibility for the old directory layout. It also allows one to
+keep all Cargo related data in one place if one wishes to.
+
+## Windows
+
+We'll obtain each of the following directories using the correct API.
+
+```
+cache:    AppData\Local\Temp\Cargo
+config:   AppData\Roaming\Cargo
+binaries: AppData\Local\Programs\Cargo
```

Comment:
%LOCALAPPDATA%\Cargo\Programs would be better - the convention in AppData is to own the subdirectory, not to create a generic space under it and then claim a subspace under that.

Reply:
Incorrect.
`FOLDERID_UserProgramFiles` is what we want, and it defaults to `%LOCALAPPDATA%\Programs` which `AppData\Local\Programs\Cargo` falls under. Probably would be good to update the text to specify the known folder id constants specifically.

Reply:
Ah, documented https://docs.microsoft.com/en-us/windows/win32/msi/installation-context - thanks.